flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/FileSplitReadOperator.java

```diff
@@ -0,0 +1,184 @@
+		// another thread. This will change later for a cleaner design.
+		setChainingStrategy(ChainingStrategy.NEVER);
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+
+		this.format.configure(configuration);
+		this.collector = new TimestampedCollector<>(output);
+
+		TypeSerializer<OUT> serializer = typeInfo.createSerializer(getRuntimeContext().getExecutionConfig());
+		Object checkpointLock = getContainingTask().getCheckpointLock();
+
+		this.reader = new SplitReader<>(format, serializer, collector, checkpointLock);
+		this.reader.start();
+	}
+
+	@Override
+	public void processElement(StreamRecord<FileInputSplit> element) throws Exception {
+		this.reader.addSplit(element.getValue());
+	}
+
+	@Override
+	public void processWatermark(Watermark mark) throws Exception {
+		output.emitWatermark(mark);
+	}
+
+	@Override
+	public void close() throws Exception {
+		super.close();
+		this.reader.cancel();
+		this.reader.interrupt();
+		this.collector.close();
+	}
+
+	private class SplitReader<OT> extends Thread {
+
+		private boolean isRunning;
+
+		private final Queue<FileInputSplit> pendingSplits;
+
+		private final FileInputFormat<OT> format;
+		private final TypeSerializer<OT> serializer;
+
+		private final Object checkpointLock;
+		private final TimestampedCollector<OT> collector;
+
+		SplitReader(FileInputFormat<OT> format,
+					TypeSerializer<OT> serializer,
+					TimestampedCollector<OT> collector,
+					Object checkpointLock) {
+			this.format = checkNotNull(format, "Unspecified FileInputFormat.");
+			this.serializer = checkNotNull(serializer, "Unspecified Serialized.");
+
+			this.pendingSplits = new ConcurrentLinkedQueue<>();
+			this.collector = collector;
+			this.checkpointLock = checkpointLock;
+			this.isRunning = true;
+		}
+
+		void addSplit(FileInputSplit split) {
+			Preconditions.checkNotNull(split);
+			LOG.info("Adding split to read queue: " + split);
+			this.pendingSplits.add(split);
+		}
+
+		@Override
+		public void run() {
+			FileInputSplit split = null;
+
+			try {
+				while(this.isRunning) {
+
+					// get the next split to read
+					split = this.pendingSplits.poll();
+					if (split == null) {
+						Thread.sleep(50);
+						continue;
+					}
+
+					boolean isOpen = false;
+					try {
+						this.format.open(split);
+						isOpen = true;
+						OT nextElement = serializer.createInstance();
+						do {
+							nextElement = format.nextRecord(nextElement);
+							if (nextElement != null) {
+								synchronized (checkpointLock) {
+									collector.collect(nextElement);
+									// checkpointing should be here.
+								}
+							}
+						} while (nextElement != null && !format.reachedEnd());
+					} finally {
+						if (isOpen) {
+							this.format.close();
+						}
+					}
+				}
+
+				LOG.info("Split Reader terminated, and exiting normally.");
+			} catch (IOException e) {
+				throw new RuntimeException("Unable to open split: " + split, e.getCause());
```

Comment:
We should wrap this in an `AsyncException` and store it, so that `processElement` can rethrow it. Please have a look at `StreamTask.asyncException`, we can basically copy the stuff from there.

This is not completely fool-proof, since we only recognize the exception if new data arrives in `processElement` but it is better than nothing.